expr.c

```diff
@@ -0,0 +1,649 @@
+/*  expr.c -- filter expression parsing and processing.
+
+    Copyright (C) 2020 Genome Research Ltd.
+
+    Author: James Bonfield <jkb@sanger.ac.uk>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notices and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.  */
+
+// TODO:
+// - add maths functions.  pow, sqrt, log, min, max, ?
+// - ?: operator for conditionals?
+
+#include <config.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <float.h>
+#include <regex.h> // may need configure rule for this
+
+#include "expr.h"
+
+// Could also cache fexpr_t stack here for kstring reuse?
+#define MAX_REGEX 10
+struct sam_filter_t {
+    char *str;
+    int parsed;
+    int curr_regex, max_regex;
+    regex_t preg[MAX_REGEX];
+};
+
+/*
+ * This is designed to be mostly C like with mostly same the precedence rules,
+ * with the exception of bit operators (widely considered as a mistake in C).
+ * It's not full C (eg no bit-shifting), but good enough for our purposes.
+ *
+ * Supported syntax, in order of precedence:
+ *
+ * Grouping:      (, ),   eg "(1+2)*3"
+ * Values:        integers, floats, strings or variables
+ * Unary ops:     +, -, !, ~  eg -10 +10, !10 (0), ~5 (bitwise not)
+ * Math ops:      *, /, %  [TODO: add // for floor division?]
+ * Math ops:      +, -
+ * Bit-wise:      &, |, ^  [NB as 3 precedence levels, in that order]
+ * Conditionals:  >, >=, <, <=,
+ * Equality:      ==, !=, =~ !~
+ * Boolean:       &&, ||
+ */
+
+// Skip to start of term
+static char *ws(char *str) {
+    while (*str && (*str == ' ' || *str == '\t'))
+        str++;
+    return str;
+}
+
+static int expression(sam_filter_t *filt, void *data, sym_func *fn,
+                      char *str, char **end, fexpr_t *res);
+
+// Copied from htslib's textutils_internal.h.  Delete after migration
```

Comment:
You need to follow the instructions in this comment!

Reply:
Oops.  Also now done.